File: day_16_prefix_sum/day_16_prefix_sum.js
/**
 * Prefix Sum: Practical JavaScript Study File
 *
 * This file progresses from basic one-dimensional prefix sums to
 * two-dimensional queries, subarray problems, difference arrays, Fenwick
 * trees, and application-oriented examples.
 *
 * It uses only standard JavaScript features and can run in Node.js or
 * another modern JavaScript runtime.
 */

// ---------------------------------------------------------------------------
// 1. Basic prefix sum
// ---------------------------------------------------------------------------

function buildPrefixSum(values) {
    // prefix[i] stores the sum of values[0] through values[i - 1].
    // prefix[0] is deliberately zero so ranges beginning at index 0
    // require no special case.
    const prefix = new Array(values.length + 1).fill(0);

    for (let index = 0; index < values.length; index += 1) {
        prefix[index + 1] = prefix[index] + values[index];
    }

    return prefix;
}

function rangeSum(prefix, left, right) {
    if (
        left < 0 ||
        right < left ||
        right + 1 >= prefix.length
    ) {
        throw new RangeError("Invalid inclusive range");
    }

    return prefix[right + 1] - prefix[left];
}

function naiveRangeSum(values, left, right) {
    if (left < 0 || right < left || right >= values.length) {
        throw new RangeError("Invalid range");
    }

    let total = 0;

    for (let index = left; index <= right; index += 1) {
        total += values[index];
    }

    return total;
}


// ---------------------------------------------------------------------------
// 2. Static range-query class
// ---------------------------------------------------------------------------

class PrefixRangeQuery {
    constructor(values) {
        this.values = Object.freeze([...values]);
        this.prefix = buildPrefixSum(this.values);
    }

    query(left, right) {
        return rangeSum(this.prefix, left, right);
    }
}


// ---------------------------------------------------------------------------
// 3. Conditional counting
// ---------------------------------------------------------------------------

function buildPredicatePrefix(values, predicate) {
    // Convert the condition into 0/1 values, then prefix-sum them.
    const flags = values.map(value => predicate(value) ? 1 : 0);
    return buildPrefixSum(flags);
}

function demonstrateConditionalCounting() {
    const values = [3, 12, 7, 18, 5, 20, 9];
    const evenPrefix = buildPredicatePrefix(
        values,
        value => value % 2 === 0
    );

    console.log(
        "Even values in [1, 5]:",
        rangeSum(evenPrefix, 1, 5)
    );
}


// ---------------------------------------------------------------------------
// 4. Two-dimensional prefix sums
// ---------------------------------------------------------------------------

function build2DPrefix(matrix) {
    if (matrix.length === 0) {
        return [[0]];
    }

    const columns = matrix[0].length;

    if (matrix.some(row => row.length !== columns)) {
        throw new Error("Matrix rows must have equal lengths");
    }

    const prefix = Array.from(
        { length: matrix.length + 1 },
        () => new Array(columns + 1).fill(0)
    );

    for (let row = 0; row < matrix.length; row += 1) {
        for (let column = 0; column < columns; column += 1) {
            prefix[row + 1][column + 1] =
                matrix[row][column] +
                prefix[row][column + 1] +
                prefix[row + 1][column] -
                prefix[row][column];
        }
    }

    return prefix;
}

function rectangleSum(prefix, top, left, bottom, right) {
    if (top > bottom || left > right) {
        throw new RangeError("Invalid rectangle");
    }

    return (
        prefix[bottom + 1][right + 1] -
        prefix[top][right + 1] -
        prefix[bottom + 1][left] +
        prefix[top][left]
    );
}


// ---------------------------------------------------------------------------
// 5. Counting subarrays with a target sum
// ---------------------------------------------------------------------------

function countSubarraysWithSum(values, target) {
    /*
     * Let P[j] be a prefix sum.
     *
     * A subarray from i through j has sum target when:
     *
     *     P[j + 1] - P[i] = target
     *
     * Therefore:
     *
     *     P[i] = P[j + 1] - target
     *
     * A Map stores how many times each previous prefix sum occurred.
     */
    const frequencies = new Map([[0, 1]]);
    let currentSum = 0;
    let answer = 0;

    for (const value of values) {
        currentSum += value;

        answer += frequencies.get(currentSum - target) ?? 0;

        frequencies.set(
            currentSum,
            (frequencies.get(currentSum) ?? 0) + 1
        );
    }

    return answer;
}


// ---------------------------------------------------------------------------
// 6. Longest subarray with a target sum
// ---------------------------------------------------------------------------

function longestSubarrayWithSum(values, target) {
    // Store the earliest index for each prefix sum.
    // The earliest occurrence gives the longest possible interval later.
    const firstSeen = new Map([[0, -1]]);
    let currentSum = 0;
    let best = null;

    for (let index = 0; index < values.length; index += 1) {
        currentSum += values[index];

        const needed = currentSum - target;

        if (firstSeen.has(needed)) {
            const candidateLeft = firstSeen.get(needed) + 1;
            const candidateRight = index;

            if (
                best === null ||
                candidateRight - candidateLeft >
                best.right - best.left
            ) {
                best = {
                    left: candidateLeft,
                    right: candidateRight
                };
            }
        }

        if (!firstSeen.has(currentSum)) {
            firstSeen.set(currentSum, index);
        }
    }

    return best;
}


// ---------------------------------------------------------------------------
// 7. Difference array for repeated range updates
// ---------------------------------------------------------------------------

function applyRangeAdditions(size, operations) {
    /*
     * Instead of modifying every element in [left, right], record two
     * boundary events. A final prefix sum turns these events into values.
     */
    const difference = new Array(size + 1).fill(0);

    for (const [left, right, amount] of operations) {
        if (
            left < 0 ||
            right < left ||
            right >= size
        ) {
            throw new RangeError("Invalid update range");
        }

        difference[left] += amount;
        difference[right + 1] -= amount;
    }

    const result = new Array(size);
    let running = 0;

    for (let index = 0; index < size; index += 1) {
        running += difference[index];
        result[index] = running;
    }

    return result;
}


// ---------------------------------------------------------------------------
// 8. Prefix XOR
// ---------------------------------------------------------------------------

function buildPrefixXOR(values) {
    const prefix = new Array(values.length + 1).fill(0);

    for (let index = 0; index < values.length; index += 1) {
        prefix[index + 1] = prefix[index] ^ values[index];
    }

    return prefix;
}

function xorRangeQuery(prefix, left, right) {
    // XOR is self-inverse: x ^ x = 0.
    return prefix[right + 1] ^ prefix[left];
}


// ---------------------------------------------------------------------------
// 9. Fenwick tree
// ---------------------------------------------------------------------------

class FenwickTree {
    constructor(values) {
        this.size = values.length;
        this.tree = new Array(this.size + 1).fill(0);

        for (let index = 0; index < values.length; index += 1) {
            this.add(index + 1, values[index]);
        }
    }

    add(oneBasedIndex, delta) {
        if (
            oneBasedIndex < 1 ||
            oneBasedIndex > this.size
        ) {
            throw new RangeError("Fenwick index out of range");
        }

        let index = oneBasedIndex;

        while (index <= this.size) {
            this.tree[index] += delta;
            index += index & -index;
        }
    }

    prefixSum(count) {
        if (count < 0 || count > this.size) {
            throw new RangeError("Fenwick prefix length out of range");
        }

        let index = count;
        let total = 0;

        while (index > 0) {
            total += this.tree[index];
            index -= index & -index;
        }

        return total;
    }

    rangeSum(left, right) {
        if (
            left < 0 ||
            right < left ||
            right >= this.size
        ) {
            throw new RangeError("Invalid range");
        }

        return (
            this.prefixSum(right + 1) -
            this.prefixSum(left)
        );
    }
}


// ---------------------------------------------------------------------------
// 10. Segment tree
// ---------------------------------------------------------------------------

class SegmentTree {
    constructor(values) {
        this.size = 1;

        while (this.size < values.length) {
            this.size *= 2;
        }

        this.tree = new Array(this.size * 2).fill(0);

        for (let index = 0; index < values.length; index += 1) {
            this.tree[this.size + index] = values[index];
        }

        for (
            let index = this.size - 1;
            index > 0;
            index -= 1
        ) {
            this.tree[index] =
                this.tree[index * 2] +
                this.tree[index * 2 + 1];
        }
    }

    update(index, value) {
        if (index < 0 || index >= this.size) {
            throw new RangeError("Segment-tree index out of range");
        }

        let position = this.size + index;
        this.tree[position] = value;

        while (position > 1) {
            position = Math.floor(position / 2);

            this.tree[position] =
                this.tree[position * 2] +
                this.tree[position * 2 + 1];
        }
    }

    query(left, right) {
        if (
            left < 0 ||
            right < left ||
            right >= this.size
        ) {
            throw new RangeError("Invalid range");
        }

        let l = this.size + left;
        let r = this.size + right;
        let result = 0;

        while (l <= r) {
            if (l % 2 === 1) {
                result += this.tree[l];
                l += 1;
            }

            if (r % 2 === 0) {
                result += this.tree[r];
                r -= 1;
            }

            l = Math.floor(l / 2);
            r = Math.floor(r / 2);
        }

        return result;
    }
}


// ---------------------------------------------------------------------------
// 11. Exact arithmetic note for JavaScript
// ---------------------------------------------------------------------------

function demonstrateNumberLimitations() {
    /*
     * JavaScript's ordinary Number is IEEE-754 double precision.
     * Integer arithmetic is exact only through Number.MAX_SAFE_INTEGER.
     *
     * BigInt is appropriate when prefix sums can exceed that exact range.
     */
    const safeValues = [
        Number.MAX_SAFE_INTEGER - 2,
        1,
        1
    ];

    const safePrefix = buildPrefixSum(safeValues);

    console.log(
        "Safe integer prefix:",
        safePrefix
    );

    const bigValues = [
        9007199254740991n,
        10n,
        20n
    ];

    const bigPrefix = [0n];

    for (const value of bigValues) {
        bigPrefix.push(bigPrefix.at(-1) + value);
    }

    console.log(
        "BigInt prefix:",
        bigPrefix
    );
}


// ---------------------------------------------------------------------------
// 12. Performance comparison
// ---------------------------------------------------------------------------

function demonstratePerformance() {
    const values = Array.from(
        { length: 20_000 },
        (_, index) => (index * 17) % 101 - 50
    );

    const queries = Array.from(
        { length: 5_000 },
        (_, index) => {
            const left = (index * 37) % values.length;
            const right =
                Math.min(
                    values.length - 1,
                    left + ((index * 13) % 500)
                );

            return [left, right];
        }
    );

    const startNaive = performance.now();
    let naiveTotal = 0;

    for (const [left, right] of queries) {
        naiveTotal += naiveRangeSum(values, left, right);
    }

    const naiveTime = performance.now() - startNaive;

    const startPrefix = performance.now();
    const prefix = buildPrefixSum(values);
    let prefixTotal = 0;

    for (const [left, right] of queries) {
        prefixTotal += rangeSum(prefix, left, right);
    }

    const prefixTime = performance.now() - startPrefix;

    if (naiveTotal !== prefixTotal) {
        throw new Error("Performance comparison produced different results");
    }

    console.log(
        `Naive: ${naiveTime.toFixed(3)} ms; ` +
        `Prefix: ${prefixTime.toFixed(3)} ms`
    );
}


// ---------------------------------------------------------------------------
// 13. Correctness tests
// ---------------------------------------------------------------------------

function assertEqual(actual, expected, message) {
    if (actual !== expected) {
        throw new Error(
            `${message}: expected ${expected}, received ${actual}`
        );
    }
}

function runTests() {
    const values = [4, 2, 7, 1, 5, 3];
    const prefix = buildPrefixSum(values);

    assertEqual(
        rangeSum(prefix, 1, 4),
        15,
        "Basic range sum"
    );

    assertEqual(
        naiveRangeSum(values, 1, 4),
        15,
        "Naive range sum"
    );

    const matrix = [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]
    ];

    const matrixPrefix = build2DPrefix(matrix);

    assertEqual(
        rectangleSum(matrixPrefix, 1, 1, 2, 2),
        28,
        "Rectangle sum"
    );

    assertEqual(
        countSubarraysWithSum([1, 2, 1, 2, 1], 3),
        4,
        "Subarray count"
    );

    assertEqual(
        applyRangeAdditions(
            6,
            [[1, 3, 5], [2, 5, 2], [0, 1, 4]]
        ).join(","),
        "4,9,7,7,2,2",
        "Difference array"
    );

    const fenwick = new FenwickTree([2, 4, 6, 8, 10]);

    assertEqual(
        fenwick.rangeSum(1, 3),
        18,
        "Fenwick query"
    );

    fenwick.add(3, 5);

    assertEqual(
        fenwick.rangeSum(1, 3),
        23,
        "Fenwick update"
    );

    const segment = new SegmentTree([2, 4, 6, 8, 10]);

    assertEqual(
        segment.query(1, 3),
        18,
        "Segment query"
    );

    segment.update(2, 11);

    assertEqual(
        segment.query(1, 3),
        23,
        "Segment update"
    );

    console.log("All JavaScript tests passed.");
}


// ---------------------------------------------------------------------------
// 14. Main demonstration
// ---------------------------------------------------------------------------

function main() {
    console.log("PREFIX SUM STUDY PROGRAM");
    console.log("========================");

    const values = [4, 2, 7, 1, 5, 3];
    const prefix = buildPrefixSum(values);

    console.log("Values:", values);
    console.log("Prefix:", prefix);
    console.log("Sum [1, 4]:", rangeSum(prefix, 1, 4));

    demonstrateConditionalCounting();

    const matrix = [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12]
    ];

    const matrixPrefix = build2DPrefix(matrix);

    console.log(
        "2D rectangle [0..1][1..3]:",
        rectangleSum(matrixPrefix, 0, 1, 1, 3)
    );

    console.log(
        "Subarrays with sum 3:",
        countSubarraysWithSum([1, 2, 1, 2, 1], 3)
    );

    console.log(
        "Longest target-sum subarray:",
        longestSubarrayWithSum([1, -1, 5, -2, 3], 3)
    );

    console.log(
        "Range additions:",
        applyRangeAdditions(
            6,
            [[1, 3, 5], [2, 5, 2], [0, 1, 4]]
        )
    );

    const xorValues = [5, 2, 7, 3, 9];
    const xorPrefix = buildPrefixXOR(xorValues);

    console.log(
        "Prefix XOR [1, 3]:",
        xorRangeQuery(xorPrefix, 1, 3)
    );

    const queryStructure = new PrefixRangeQuery([10, 20, 30, 40, 50]);

    console.log(
        "Immutable query [1, 3]:",
        queryStructure.query(1, 3)
    );

    const fenwick = new FenwickTree([2, 4, 6, 8, 10]);

    console.log(
        "Fenwick [1, 3]:",
        fenwick.rangeSum(1, 3)
    );

    demonstrateNumberLimitations();
    demonstratePerformance();
    runTests();
}

main();

File: day_16_prefix_sum/day_16_prefix_sum.cpp
/*
    Prefix Sum: Industry-Style Telemetry Analytics Case Study

    Scenario
    --------
    A monitoring service receives a sequence of integer measurements from
    infrastructure sensors. Analysts repeatedly request statistics over
    time intervals, while an administrative component occasionally changes
    individual measurements.

    The implementation begins with direct scanning, evolves to immutable
    prefix sums for high-volume read workloads, then uses a Fenwick tree
    when point updates and range queries must coexist.

    Standard: C++17
*/

#include <algorithm>
#include <cassert>
#include <chrono>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <limits>
#include <numeric>
#include <optional>
#include <random>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>


// ---------------------------------------------------------------------------
// Domain model
// ---------------------------------------------------------------------------

struct SensorReading {
    std::string sensor_id;
    std::vector<std::int64_t> measurements;
};


// ---------------------------------------------------------------------------
// Validation utilities
// ---------------------------------------------------------------------------

void validate_range(
    std::size_t size,
    std::size_t left,
    std::size_t right
) {
    if (size == 0 || left > right || right >= size) {
        throw std::out_of_range("Invalid inclusive range");
    }
}


// ---------------------------------------------------------------------------
// Direct implementation
// ---------------------------------------------------------------------------

class DirectRangeCalculator {
public:
    explicit DirectRangeCalculator(
        const std::vector<std::int64_t>& values
    )
        : values_(values) {}

    std::int64_t range_sum(
        std::size_t left,
        std::size_t right
    ) const {
        validate_range(values_.size(), left, right);

        std::int64_t result = 0;

        for (std::size_t index = left; index <= right; ++index) {
            result += values_[index];
        }

        return result;
    }

private:
    const std::vector<std::int64_t>& values_;
};


// ---------------------------------------------------------------------------
// Immutable prefix-sum implementation
// ---------------------------------------------------------------------------

class PrefixSumArray {
public:
    explicit PrefixSumArray(
        const std::vector<std::int64_t>& values
    ) {
        prefix_.resize(values.size() + 1, 0);

        for (std::size_t index = 0; index < values.size(); ++index) {
            prefix_[index + 1] =
                prefix_[index] + values[index];
        }
    }

    std::int64_t range_sum(
        std::size_t left,
        std::size_t right
    ) const {
        const std::size_t value_count = prefix_.size() - 1;

        validate_range(value_count, left, right);

        return prefix_[right + 1] - prefix_[left];
    }

    std::int64_t total_sum() const {
        return prefix_.back();
    }

    std::size_t size() const {
        return prefix_.size() - 1;
    }

private:
    std::vector<std::int64_t> prefix_;
};


// ---------------------------------------------------------------------------
// Extended prefix statistics
// ---------------------------------------------------------------------------

class PrefixStatistics {
public:
    explicit PrefixStatistics(
        const std::vector<std::int64_t>& values
    ) {
        count_.resize(values.size() + 1, 0);
        sum_.resize(values.size() + 1, 0);
        squares_.resize(values.size() + 1, 0);

        for (std::size_t index = 0; index < values.size(); ++index) {
            const auto value = values[index];

            count_[index + 1] = count_[index] + 1;
            sum_[index + 1] = sum_[index] + value;

            // This assumes the application has chosen an integer range
            // where squaring is safe. Production systems should define
            // explicit overflow policy based on domain limits.
            squares_[index + 1] =
                squares_[index] + value * value;
        }
    }

    std::int64_t sum(
        std::size_t left,
        std::size_t right
    ) const {
        validate_range(count_.size() - 1, left, right);
        return sum_[right + 1] - sum_[left];
    }

    std::size_t count(
        std::size_t left,
        std::size_t right
    ) const {
        validate_range(count_.size() - 1, left, right);
        return static_cast<std::size_t>(
            count_[right + 1] - count_[left]
        );
    }

    double mean(
        std::size_t left,
        std::size_t right
    ) const {
        return static_cast<double>(sum(left, right)) /
               static_cast<double>(count(left, right));
    }

    double variance(
        std::size_t left,
        std::size_t right
    ) const {
        const auto n = static_cast<double>(count(left, right));
        const auto total = static_cast<double>(sum(left, right));
        const auto square_total = static_cast<double>(
            squares_[right + 1] - squares_[left]
        );

        const double average = total / n;

        // Population variance:
        // E[X^2] - E[X]^2
        return square_total / n - average * average;
    }

private:
    std::vector<std::int64_t> count_;
    std::vector<std::int64_t> sum_;
    std::vector<std::int64_t> squares_;
};


// ---------------------------------------------------------------------------
// Two-dimensional prefix sums
// ---------------------------------------------------------------------------

class MatrixPrefixSum {
public:
    explicit MatrixPrefixSum(
        const std::vector<std::vector<std::int64_t>>& matrix
    ) {
        if (matrix.empty()) {
            prefix_ = {{0}};
            return;
        }

        const std::size_t columns = matrix.front().size();

        for (const auto& row : matrix) {
            if (row.size() != columns) {
                throw std::invalid_argument(
                    "Matrix must be rectangular"
                );
            }
        }

        prefix_.assign(
            matrix.size() + 1,
            std::vector<std::int64_t>(columns + 1, 0)
        );

        for (std::size_t row = 0; row < matrix.size(); ++row) {
            for (std::size_t column = 0; column < columns; ++column) {
                prefix_[row + 1][column + 1] =
                    matrix[row][column]
                    + prefix_[row][column + 1]
                    + prefix_[row + 1][column]
                    - prefix_[row][column];
            }
        }
    }

    std::int64_t rectangle_sum(
        std::size_t top,
        std::size_t left,
        std::size_t bottom,
        std::size_t right
    ) const {
        if (
            top > bottom ||
            left > right ||
            bottom + 1 >= prefix_.size() ||
            right + 1 >= prefix_.front().size()
        ) {
            throw std::out_of_range("Invalid rectangle");
        }

        // Inclusion-exclusion:
        //
        // complete rectangle
        // - region above
        // - region left
        // + upper-left overlap
        return
            prefix_[bottom + 1][right + 1]
            - prefix_[top][right + 1]
            - prefix_[bottom + 1][left]
            + prefix_[top][left];
    }

private:
    std::vector<std::vector<std::int64_t>> prefix_;
};


// ---------------------------------------------------------------------------
// Difference array
// ---------------------------------------------------------------------------

class RangeUpdateAccumulator {
public:
    explicit RangeUpdateAccumulator(std::size_t size)
        : difference_(size + 1, 0) {}

    void add(
        std::size_t left,
        std::size_t right,
        std::int64_t amount
    ) {
        const std::size_t size = difference_.size() - 1;
        validate_range(size, left, right);

        difference_[left] += amount;
        difference_[right + 1] -= amount;
    }

    std::vector<std::int64_t> materialize() const {
        std::vector<std::int64_t> values(
            difference_.size() - 1
        );

        std::int64_t running = 0;

        for (std::size_t index = 0; index < values.size(); ++index) {
            running += difference_[index];
            values[index] = running;
        }

        return values;
    }

private:
    std::vector<std::int64_t> difference_;
};


// ---------------------------------------------------------------------------
// Fenwick tree for dynamic measurements
// ---------------------------------------------------------------------------

class FenwickTree {
public:
    explicit FenwickTree(
        const std::vector<std::int64_t>& values
    )
        : tree_(values.size() + 1, 0) {
        for (std::size_t index = 0; index < values.size(); ++index) {
            add(index + 1, values[index]);
        }
    }

    void add(
        std::size_t one_based_index,
        std::int64_t delta
    ) {
        if (
            one_based_index == 0 ||
            one_based_index >= tree_.size()
        ) {
            throw std::out_of_range(
                "Fenwick index out of range"
            );
        }

        std::size_t index = one_based_index;

        while (index < tree_.size()) {
            tree_[index] += delta;
            index += index & (~index + 1);
        }
    }

    std::int64_t prefix_sum(std::size_t count) const {
        if (count >= tree_.size()) {
            throw std::out_of_range(
                "Fenwick prefix length out of range"
            );
        }

        std::int64_t result = 0;
        std::size_t index = count;

        while (index > 0) {
            result += tree_[index];
            index &= index - 1;
        }

        return result;
    }

    std::int64_t range_sum(
        std::size_t left,
        std::size_t right
    ) const {
        const std::size_t size = tree_.size() - 1;
        validate_range(size, left, right);

        return prefix_sum(right + 1) - prefix_sum(left);
    }

private:
    std::vector<std::int64_t> tree_;
};


// ---------------------------------------------------------------------------
// Analytics service
// ---------------------------------------------------------------------------

class TelemetryAnalyticsService {
public:
    explicit TelemetryAnalyticsService(
        SensorReading reading
    )
        : reading_(std::move(reading)),
          prefix_(reading_.measurements),
          statistics_(reading_.measurements),
          dynamic_tree_(reading_.measurements) {}

    const std::string& sensor_id() const {
        return reading_.sensor_id;
    }

    std::int64_t interval_sum(
        std::size_t start,
        std::size_t end
    ) const {
        return prefix_.range_sum(start, end);
    }

    double interval_mean(
        std::size_t start,
        std::size_t end
    ) const {
        return statistics_.mean(start, end);
    }

    double interval_variance(
        std::size_t start,
        std::size_t end
    ) const {
        return statistics_.variance(start, end);
    }

    std::int64_t dynamic_interval_sum(
        std::size_t start,
        std::size_t end
    ) const {
        return dynamic_tree_.range_sum(start, end);
    }

    void update_measurement(
        std::size_t index,
        std::int64_t new_value
    ) {
        if (index >= reading_.measurements.size()) {
            throw std::out_of_range(
                "Measurement index out of range"
            );
        }

        const std::int64_t old_value =
            reading_.measurements[index];

        const std::int64_t delta = new_value - old_value;

        reading_.measurements[index] = new_value;

        // The immutable prefix and statistics structures represent the
        // original snapshot. The Fenwick tree is the dynamic structure.
        dynamic_tree_.add(index + 1, delta);
    }

private:
    SensorReading reading_;
    PrefixSumArray prefix_;
    PrefixStatistics statistics_;
    FenwickTree dynamic_tree_;
};


// ---------------------------------------------------------------------------
// Subarray analytics
// ---------------------------------------------------------------------------

std::size_t count_subarrays_with_sum(
    const std::vector<std::int64_t>& values,
    std::int64_t target
) {
    /*
        If:
            prefix[j] - prefix[i] = target

        then:
            prefix[i] = prefix[j] - target

        An unordered_map is commonly used for the frequency table.
    */

    // Included here through a local standard-library associative structure.
    std::unordered_map<std::int64_t, std::size_t> frequency;

    frequency[0] = 1;

    std::int64_t current = 0;
    std::size_t answer = 0;

    for (const auto value : values) {
        current += value;

        const auto found = frequency.find(current - target);

        if (found != frequency.end()) {
            answer += found->second;
        }

        ++frequency[current];
    }

    return answer;
}


// ---------------------------------------------------------------------------
// Brute-force validation for tests
// ---------------------------------------------------------------------------

std::int64_t brute_force_sum(
    const std::vector<std::int64_t>& values,
    std::size_t left,
    std::size_t right
) {
    validate_range(values.size(), left, right);

    return std::accumulate(
        values.begin() + static_cast<std::ptrdiff_t>(left),
        values.begin() + static_cast<std::ptrdiff_t>(right + 1),
        std::int64_t{0}
    );
}


// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

void run_tests() {
    const std::vector<std::int64_t> values{
        4, 2, 7, 1, 5, 3
    };

    PrefixSumArray prefix(values);

    for (std::size_t left = 0; left < values.size(); ++left) {
        for (
            std::size_t right = left;
            right < values.size();
            ++right
        ) {
            assert(
                prefix.range_sum(left, right)
                == brute_force_sum(values, left, right)
            );
        }
    }

    MatrixPrefixSum matrix({
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    });

    assert(matrix.rectangle_sum(1, 1, 2, 2) == 28);

    RangeUpdateAccumulator updates(6);

    updates.add(1, 3, 5);
    updates.add(2, 5, 2);
    updates.add(0, 1, 4);

    const auto result = updates.materialize();

    assert(
        result == std::vector<std::int64_t>{
            4, 9, 7, 7, 2, 2
        }
    );

    FenwickTree tree(values);

    assert(tree.range_sum(1, 4) == 18);

    tree.add(3, 5);

    assert(tree.range_sum(1, 4) == 23);

    assert(
        count_subarrays_with_sum(
            {1, 2, 1, 2, 1},
            3
        ) == 4
    );
}


// ---------------------------------------------------------------------------
// Demonstration
// ---------------------------------------------------------------------------

void print_vector(
    const std::vector<std::int64_t>& values
) {
    std::cout << "[";

    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index != 0) {
            std::cout << ", ";
        }

        std::cout << values[index];
    }

    std::cout << "]";
}


int main() {
    try {
        run_tests();

        std::cout << "PREFIX SUM TELEMETRY CASE STUDY\n";
        std::cout << "================================\n\n";

        SensorReading reading{
            "SENSOR-ALPHA",
            {12, 15, 11, 20, 18, 14, 17, 21, 13, 16}
        };

        TelemetryAnalyticsService service(std::move(reading));

        std::cout << "Sensor: "
                  << service.sensor_id()
                  << "\n";

        std::cout << "Static interval sum [2, 7]: "
                  << service.interval_sum(2, 7)
                  << "\n";

        std::cout << "Static interval mean [2, 7]: "
                  << std::fixed
                  << std::setprecision(2)
                  << service.interval_mean(2, 7)
                  << "\n";

        std::cout << "Static interval variance [2, 7]: "
                  << service.interval_variance(2, 7)
                  << "\n";

        std::cout << "Dynamic interval sum [2, 7]: "
                  << service.dynamic_interval_sum(2, 7)
                  << "\n";

        /*
            An operational correction arrives for measurement index 4.

            Prefix sums are excellent for immutable data but do not make
            arbitrary point updates O(1). The Fenwick tree handles the
            corrected value without rebuilding the complete prefix array.
        */
        service.update_measurement(4, 25);

        std::cout << "After correcting index 4 to 25:\n";

        std::cout << "Dynamic interval sum [2, 7]: "
                  << service.dynamic_interval_sum(2, 7)
                  << "\n\n";

        MatrixPrefixSum regional_matrix({
            {10, 12, 15, 11},
            {14, 13, 18, 20},
            {16, 19, 17, 21},
            {12, 14, 22, 25}
        });

        std::cout << "Regional rectangle [1..3][1..2]: "
                  << regional_matrix.rectangle_sum(
                         1, 1, 3, 2
                     )
                  << "\n";

        RangeUpdateAccumulator maintenance_schedule(8);

        maintenance_schedule.add(1, 4, 3);
        maintenance_schedule.add(3, 6, 5);
        maintenance_schedule.add(0, 2, 2);

        std::cout << "Aggregated maintenance load: ";

        const auto schedule =
            maintenance_schedule.materialize();

        print_vector(schedule);

        std::cout << "\n";

        std::cout << "Subarrays summing to 20: "
                  << count_subarrays_with_sum(
                         {5, 15, 10, -5, 20, -10, 30},
                         20
                     )
                  << "\n";

        std::cout << "\nAll C++ correctness tests passed.\n";

        /*
            Complexity:
              Prefix construction: O(n)
              Static range query: O(1)
              Fenwick point update: O(log n)
              Fenwick range query: O(log n)
              2D prefix construction: O(rows * columns)
              2D rectangle query: O(1)
              Difference-array update: O(1)
              Difference-array materialization: O(n)

            Production considerations:
              - Establish whether signed overflow is possible.
              - Use suitable integer widths for the measurement domain.
              - Validate externally supplied indices.
              - Treat immutable snapshots and live data as separate concepts.
              - Avoid rebuilding large prefix arrays for every small update.
              - Benchmark using workload characteristics rather than isolated
                microbenchmarks.
        */

        return 0;
    }
    catch (const std::exception& error) {
        std::cerr << "Application error: "
                  << error.what()
                  << "\n";
        return 1;
    }
}
